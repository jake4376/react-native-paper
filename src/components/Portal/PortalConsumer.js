/* @flow */
/* eslint-disable react/no-unused-prop-types */

import * as React from 'react';
import type { PortalMethods } from './PortalHost';
import type { PortalProps } from './Portal';

type Props = {
  manager: PortalMethods,
  props: PortalProps,
};

export default class PortalConsumer extends React.Component<Props> {
  componentWillMount() {
    this._key = this.props.manager.mount(this.props.props);
  }

  componentWillReceiveProps(nextProps: Props) {
    this.props.manager.update(this._key, nextProps.props);
  }

  componentWillUnmount() {
    this.props.manager.unmount(this._key);
  }

  _key: any;

  render() {
    return null;
  }
}
