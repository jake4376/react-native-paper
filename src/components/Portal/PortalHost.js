/* @flow */

import * as React from 'react';
import PropTypes from 'prop-types';
import { View, StyleSheet } from 'react-native';
import type { PortalProps } from './Portal';

type Props = {
  children: React.Node,
  style?: any,
};

type State = {
  portals: Array<{
    key: number,
    props: PortalProps,
  }>,
};

export const manager = 'react-native-paper$portal-manager';

/**
 * Portal host is the component which actually renders all Portals.
 */
export default class Portals extends React.Component<Props, State> {
  static childContextTypes = {
    [manager]: PropTypes.object,
  };

  state = {
    portals: [],
  };

  getChildContext() {
    return {
      [manager]: {
        mount: this._mountPortal,
        unmount: this._unmountPortal,
        update: this._updatePortal,
      },
    };
  }

  _nextId = 0;

  _mountPortal = (props: PortalProps) => {
    const key = this._nextId++;
    this.setState(state => ({
      portals: [...state.portals, { key, props }],
    }));
    return key;
  };

  _unmountPortal = (key: number) =>
    this.setState(state => ({
      portals: state.portals.filter(item => item.key !== key),
    }));

  _updatePortal = (key: number, props: PortalProps) =>
    this.setState(state => ({
      portals: state.portals.map(item => {
        if (item.key === key) {
          return { ...item, props };
        }
        return item;
      }),
    }));

  render() {
    const { portals } = this.state;
    return (
      <View {...this.props} style={[styles.container, this.props.style]}>
        {this.props.children}
        {portals
          .reduce((acc, curr) => {
            const { position = 0, children } = curr.props;
            let group = acc.find(it => it.position === position);
            if (group) {
              group = {
                position,
                items: group.items.concat([children]),
              };
              return acc.map(g => {
                if (group && g.position === position) {
                  return group;
                }
                return g;
              });
            } else {
              group = { position, items: [children] };
              return [...acc, group];
            }
          }, [])
          .map(({ position, items }) => (
            <View key={position} style={StyleSheet.absoluteFill}>
              {items}
            </View>
          ))}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});
