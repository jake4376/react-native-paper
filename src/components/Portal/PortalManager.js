/* @flow */

import * as React from 'react';
import { View, StyleSheet } from 'react-native';
import type { PortalProps } from './Portal';

type State = {
  portals: Array<{
    key: number,
    props: PortalProps,
  }>,
};

/**
 * Portal host is the component which actually renders all Portals.
 */
export default class PortalManager extends React.PureComponent<{}, State> {
  state = {
    portals: [],
  };

  mount = (key: number, props: PortalProps) => {
    this.setState(state => ({
      portals: [...state.portals, { key, props }],
    }));
  };

  update = (key: number, props: PortalProps) =>
    this.setState(state => ({
      portals: state.portals.map(item => {
        if (item.key === key) {
          return { ...item, props };
        }
        return item;
      }),
    }));

  unmount = (key: number) =>
    this.setState(state => ({
      portals: state.portals.filter(item => item.key !== key),
    }));

  render() {
    return this.state.portals
      .reduce((acc, curr) => {
        const { position = 0, children } = curr.props;
        let group = acc.find(it => it.position === position);
        if (group) {
          group = {
            position,
            items: group.items.concat([children]),
          };
          return acc.map(g => {
            if (group && g.position === position) {
              return group;
            }
            return g;
          });
        }
        group = { position, items: [children] };
        return [...acc, group];
      }, [])
      .map(({ position, items }) => (
        <View
          key={position}
          pointerEvents="box-none"
          style={StyleSheet.absoluteFill}
        >
          {items}
        </View>
      ));
  }
}
