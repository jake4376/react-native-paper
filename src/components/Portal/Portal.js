/* @flow */

import * as React from 'react';
import PropTypes from 'prop-types';
import { manager } from './PortalHost';

export type PortalProps = {
  /**
   * Position of the element in the z-axis
   */
  position?: number,
  /**
   * Content of the `Portal`.
   */
  children: React.Node,
};

type Props = PortalProps;

/**
 * Portal allows to render a component at a different place in the parent tree.
 */
export default class Portal extends React.Component<Props> {
  static contextTypes = {
    [manager]: PropTypes.object,
  };

  componentWillMount() {
    if (
      typeof this.context[manager] !== 'object' ||
      this.context[manager] === null
    ) {
      throw new Error(
        "Couldn't find portal manager in the context or props. " +
          "You need to wrap your root component in '<PortalHost />'"
      );
    }
    this._key = this.context[manager].mount(this.props);
  }

  componentWillReceiveProps(nextProps: Props) {
    this.context[manager].update(this._key, nextProps);
  }

  componentWillUnmount() {
    this.context[manager].unmount(this._key);
  }

  _key: any;

  render() {
    return null;
  }
}
