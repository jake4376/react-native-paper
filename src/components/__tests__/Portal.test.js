/* @flow */

import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Text } from 'react-native';
import PortalHost from '../Portal/PortalHost';
import Portal from '../Portal/Portal';

it('renders portal with siblings', () => {
  const wrapper = shallow(
    <PortalHost>
      <Text>Outside content</Text>
      <Portal>
        <Text>Portal content</Text>
      </Portal>
    </PortalHost>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});
