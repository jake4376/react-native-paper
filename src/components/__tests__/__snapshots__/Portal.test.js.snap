// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders portal with siblings 1`] = `
Array [
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Outside content
  </Text>,
  <View
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      Portal content
    </Text>
  </View>,
]
`;
